Add Jenkins 2.222.1 changelog (#2990)
jenkins-infra/jenkins.io: content/_data/upgrades/2-221-1.adoc

==== Header and Breadcrumb Layout Improvements

Instances with plugins that depend on specific details of the Jenkins layout (like the https://plugins.jenkins.io/simple-theme-plugin/[Simple Theme plugin]) may have user interface issues with this release.

==== Security Defaults

Global security settings no longer have a user interface setting that will disable them.
Previously, administrators could disable security with a single checkbox.
They could also disable cross-site request forgery (CSRF) protection with a single checkbox.

> Those two settings **cannot be disabled** in this release.

Reviewer: Maybe worth mentioning that some abandoned, deprecated, or proprietary plugins may still not be compatible with CSRF protection and would need to be fixed.

> Outdated agent protocols that had previously been deprecated have been removed in this release.

Author: Moved to a separate section in b687aefee5d2090b591710001b9fef34c2f1e640.  Reviewed the phrasing and details of the section with Jeff Thompson.  My first draft before review with Jeff was full of errors.  b687aefee5d2090b591710001b9fef34c2f1e640 should be accurate.

`Inbound TCP Agent Protocols` versions 1, 2, and 3 have been removed.
The `agent.jar` and `slave.jar` files on agents must be updated to the `agent.jar` version shipped with this release.

> The latest `agent.jar` can be downloaded from \https://yourserver.example.com/jnlpJars/agent.jar .

Author: Included modified text from the remoting section of the Jenkins 2.121.1 upgrade guide.  Describes minimum remoting version and the agents.  Will add reference to https://github.com/jenkinsci/windows-slave-installer-module#upgrading-old-agents

> The latest `agent.jar` can be downloaded from \https://yourserver.example.com/jnlpJars/agent.jar .

Author: Linked to windows slave installer module in b36b72a0e58fcef71670c3ab951a97b0622001e9

Password auto-fill is only allowed on the login password field.
Password fields which are **not** the login password field will no longer allow password auto-fill.
This avoids browser prompts to automatically fill password fields that are not the login password field.
Password auto-fill for other password fields can be enabled with the system property:

[source,java]
----
hudson.Functions.hidingPasswordFields=false
----

==== Security Deprecations

> The `Overall/RunScripts`, `Overall/UploadPlugins`, and `Overall/ConfigureUpdateCenter` permissions were disabled by default in 2017.

Reviewer 2: Link advisories explaining the reason?

> Those permissions are now **deprecated**.

Reviewer 2: ```suggestion
Those permissions are now **deprecated** and no longer used in the Jenkins core.
```

> Custom authorization strategy plugins that that grant `Overall/Administer` without implying oneor more of those permissions will no longer work as expected.

Reviewer 2: ```suggestion
Security configurations which grant `Overall/RunScripts`, `Overall/UploadPlugins`, and `Overall/ConfigureUpdateCenter` while not granting `Overall/Administer` will not work as expected for the Jenkins core functionality.
Plugins relying on these permissions will also eventually move to checking `Overall/Administer` instead.
Custom authorization strategies that that grant `Overall/Administer` without implying one or more of those permissions will no longer work as expected.
```

> ==== Experimental User Interface Changes

Reviewer 2: ```suggestion
==== New Experimental User Interface
```

The experimental Jenkins user interface refresh can be enabled by setting the property `jenkins.ui.refresh=true`.
The user interface refresh enables a new header color scheme.

> Additional user interface changes will be enabled in future releases with the `jenkins.ui.refresh=true` property.

Reviewer 2: This entry has low value without a link to a full story and to ways to provide feedback

==== Experimental WebSocket Protocol

Jenkins agents have commonly been connected through a dedicated network port.
The dedicated network port required additional firewall and reverse proxy configuration.
The requirement for an extra port sometimes interfered with user port limitations.

The WebSocket protocol has been added to this release as an experimental agent protocol.
WebSocket protocol allows agents to connect to Jenkins over the http and https ports, without requiring additional ports.
Refer to the link:/blog/2020/02/02/web-socket/[WebSocket] blog post for more details.

> ==== Experimental Security Additions

Reviewer 2: ```suggestion
==== Experimental SystemRead and Manager permissions
```

==== Header and Breadcrumb Layout Improvements

Instances with plugins that depend on specific details of the Jenkins layout (like the https://plugins.jenkins.io/simple-theme-plugin/[Simple Theme plugin]) may have user interface issues with this release.
In the case of the Simple Them Plugin, users or theme maintainers may need to update CSS and JavaScript to match the new layout structure.

==== Removed Agent Protocols

> Remoting changelogs for https://github.com/jenkinsci/remoting/releases/tag/remoting-4.2[4.2],

Reviewer 2: If you want to list all versions, IMHO it should be in the changelog, not in the upgrade guide. Here we need to provide links to items related to the upgrade entry, not to make admins to go through all changelogs which are mostly unrelated

https://github.com/jenkinsci/remoting/releases/tag/remoting-4.0.1[4.0.1],
https://github.com/jenkinsci/remoting/releases/tag/remoting-4.0[4.0],
https://github.com/jenkinsci/remoting/releases/tag/remoting-3.40[3.40],
https://github.com/jenkinsci/remoting/releases/tag/remoting-3.39[3.39].
https://github.com/jenkinsci/remoting/releases/tag/remoting-3.38[3.38], and
https://github.com/jenkinsci/remoting/releases/tag/remoting-3.37[3.37]

The Remoting library has been updated to resolve a security issue in remoting versions prior to 4.0.1 and to add support for the <<websocket-protocol,WebSocket protocol>>.

> The update removes outdated agent protocols that had previously been deprecated.

Reviewer 2: ```suggestion
The update removes outdated agent protocols that had previously been link:/blog/2017/08/11/remoting-update/[deprecated in Jenkins 2.75].
```

https://jenkins.io/blog/2017/08/11/remoting-update/ also provides an example of extended upgrade guidelines for Remoting

`Inbound TCP Agent Protocols` versions 1, 2, and 3 have been removed.

<<websocket-protocol,WebSocket protocol>> connections require agents using Remoting 4.2 or later.

> plugin:ssh-slaves[SSH Build Agents] will always use the latest `agent.jar`, but other agent types may need to have their `agent.jar` updated manually.

Reviewer 2: ```suggestion
plugin:ssh-slaves[SSH Build Agents Plugin] will always use the latest `agent.jar`, but other agent types may need to have their `agent.jar` updated manually.
```

> plugin:ssh-slaves[SSH Build Agents] will always use the latest `agent.jar`, but other agent types may need to have their `agent.jar` updated manually.

Reviewer 2: It is actually not 100% true, the plugin can be configured to use arbitrary Remoting JARs instead of the default one. But it is rather an undocumented feature at the moment

The following agent types should be updated:

* Inbound TCP Agent Protocol agents (formerly known as Java Web Start agents) with `agent.jar` / `slave.jar` on the disk should update to version 4.2 or newer.
  The latest `agent.jar` can be downloaded from any Jenkins server at `/jnlpJars/agent.jar` as in \https://your-jenkins-server.example.com/jnlpJars/agent.jar

> * plugin:swarm[Swarm Client] should update to version 3.18 or newer

Reviewer 2: ```suggestion
* plugin:swarm[Swarm Plugin Client] should update to version 3.18 or newer
```

> * plugin:swarm[Swarm Client] should update to version 3.18 or newer

Reviewer 2: Version is not correct. Any version with JNLP4 support will do the job

Author: @jeffret-b can you provide alternate phrasing here?  

I confirmed with the pom file that swarm client 3.18 is the first swarm client version to include remoting 4.2.  The prior version included remoting 3.40.  

Are the security changes after 3.40 only relevant for the server side?

> * Agent Docker packages offered by the Jenkins project (https://hub.docker.com/r/jenkins/slave/[jenkins/slave] and https://hub.docker.com/r/jenkins/jnlp-slave/[jenkins/jnlp-slave]) should update to version https://github.com/jenkinsci/docker-slave/releases/tag/4.0.1-1/[4.0.1-1] or newer

Reviewer 2: Version is not correct. Any version with JNLP4 support will do the job

Author: @jeffret-b can you help me with alternate phrasing here?  

We specifically used "should" rather than "must" in this context so that we allowed for other versions without specifying the other versions in detail.

Reviewer 3: We need to recommend that they upgrade to 4.0.1 or later so that they receive the fix for SECURITY-1659. They don't have to, but installing all security updates is recommended. That's why we chose "should". Not required, but recommended.

Reviewer 2: SECURITY-1659 is a master-side Security issue, isn't it? It is not relevant to agent upgades unless I am missing something

Reviewer 3: Hmm ... now I understand the confusion. Reading the [security advisory](https://jenkins.io/security/advisory/2020-01-29/) again I see that we weren't as clear as we should have been. @daniel-beck 

It's one of those cases of reading the thing multiple times and not seeing that it's leaving something out.

SECURITY-1660 is the master side. SECURITY-1659 is agent side, requiring code changes on the agent. They are very similar issues but in different places. The advisory content reads about the same for both and nothing in the advisory mentions updating agents or anything about Remoting.

==== Header and Breadcrumb Layout Improvements

Instances with plugins that depend on specific details of the Jenkins layout (like the https://plugins.jenkins.io/simple-theme-plugin/[Simple Theme plugin]) may have user interface issues with this release.

> In the case of the Simple Theme plugin, users or theme maintainers may need to update CSS and JavaScript to match the new layout structure.

Reviewer 4: This is by far the most irrelevant entry (theme users need to be prepared to handle breaking changes at any time), and it's the one on the top of the list. Why?

Author: Good point.  I'll move it to the bottom of the list.  Thanks for feedback!

Author: Moved to the bottom of the list in 0b6970941c5c8177a40ff4e8789a7c9dc14b115f

* Inbound TCP Agent Protocol agents (formerly known as Java Web Start agents) with `agent.jar` / `slave.jar` on the disk should update to version 4.2 or newer.
  The latest `agent.jar` can be downloaded from any Jenkins server at `/jnlpJars/agent.jar` as in \https://your-jenkins-server.example.com/jnlpJars/agent.jar.
  Users running Windows agents as a service with the Windows Service Wrapper should refer to https://github.com/jenkinsci/windows-slave-installer-module#upgrading-old-agents[Upgrading Old Agents] for information on agent upgrades
* plugin:swarm[Swarm Client] should update to version 3.18 or newer
* Agent Docker packages offered by the Jenkins project (https://hub.docker.com/r/jenkins/slave/[jenkins/slave] and https://hub.docker.com/r/jenkins/jnlp-slave/[jenkins/jnlp-slave]) should update to version https://github.com/jenkinsci/docker-slave/releases/tag/4.0.1-1/[4.0.1-1] or newer

==== Security Defaults

Some global security settings no longer have a user interface setting that will disable them.
Previously, administrators could disable security with a single checkbox.
They could also disable cross-site request forgery (CSRF) protection with a single checkbox.

> Those two settings **cannot be disabled** in this release.

Reviewer 4: This section is really misleading.

The first half is just a UI thing. In its place is "no security realm" and "anyone can do anything".

The second half is the likely most breaking change in the entire release for users who deliberately have CSRF protection disabled, and deserves a spot at the top.

Additionally, there's an escape hatch. It's unclear whether we should advertise it, but it's more likely to be necessary than the one below with the password form fields.

Password auto-fill is only allowed on the login password field.

> Password fields which are **not** the login password field will no longer allow password auto-fill.

Reviewer 4: Well, they attempt to prevent it.

This avoids browser prompts to automatically fill password fields that are not the login password field.

> Password auto-fill for other password fields can be enabled with the system property:

Reviewer 4: That's not what this is for. It's in case of unforeseen problems with the new password UI.

[source,java]
----
hudson.Functions.hidingPasswordFields=false
----

==== Security Deprecations

> The `Overall/RunScripts`, `Overall/UploadPlugins`, and `Overall/ConfigureUpdateCenter` permissions were disabled by default in 2017.

Reviewer 4: Misleading statement; they just couldn't be configured anymore in the most popular authorization strategies.

Reviewer 2: Role Strategy still allows it if a special System property is set. IIRC Matrix Auth plugin does the same. 

```suggestion
In 2017 the `Overall/RunScripts`, `Overall/UploadPlugins`, and `Overall/ConfigureUpdateCenter` were announced as potentially dangerous permissions.
 They were disabled by default in the most popular authorization strategies.
```

How about that?

Those permissions are now **deprecated** and no longer used in the Jenkins core.

> Security configurations which grant `Overall/RunScripts`, `Overall/UploadPlugins`, and `Overall/ConfigureUpdateCenter` while not granting `Overall/Administer` will not work as expected for the Jenkins core functionality.

Reviewer 4: Overly vague.

> Plugins relying on these permissions will also eventually move to checking `Overall/Administer` instead.

Reviewer 4: What does this tell an administrator?

==== Header and Breadcrumb Layout Improvements

Instances with plugins that depend on specific details of the Jenkins layout (like the https://plugins.jenkins.io/simple-theme-plugin/[Simple Theme plugin]) may have user interface issues with this release.

> In the case of the Simple Them Plugin, users or theme maintainers may need to update CSS and JavaScript to match the new layout structure.

Reviewer: ```suggestion
In the case of the Simple Theme plugin, users or theme maintainers may need to update CSS and JavaScript to match the new layout structure.
```

Reviewer 2: sorry for the typo

==== Security Defaults

> Some Global security settings no longer have a user interface setting that will disable them.

Reviewer: ```suggestion
Some global security settings no longer have a user interface setting that will disable them.
```
